subprojects/xcfa/xcfa/README.md

```diff
@@ -0,0 +1,123 @@
+* Basic struct and array support
+* Basic (and experimental) flat-memory based pointer handling
+* ILP32 and LP64 type system support (with possible extensibility)
+
+### Limitations, known bugs
+
+A number of features are either not well tested or in certain aspects buggy. Error handling is done on a best-effort level as the C specification is way too complex to handle entirely correctly. In most cases an error, or a warning message is displayed, but in some unexpected situations a silent failure is also possible.
+
+In particular, the following features are either not implemented or can produce buggy models:
+
+* Floating-point pointers (produces an exception when (de)referred)
+* Structs including arrays and arrays including structs (produces an exception when accessed)
+* Enums (only produces a warning, behaves as a normal signed int would)
+* Unions (produces an exception when any element is accessed) 
+* Alignof, Sizeof, various extensions (produces a parsing exception)
+* Pointer arithmetic (produces an exception)
+* Using a non-specific subtype of `char` (produces a warning) - use `(un)signed char` instead
+* Includes and other preprocessor directives (produces an exception or fails silently!)
+* Arrays are not pointers and pointers cannot be used as arrays (produces an exception)
+* Memory access is _not_ checked, and therefore it is easy to receive a faulty value by over-indexing an array (fails silently!)
+* Array elements are not range-checked and therefore false positives are possible (consider `char c[2]; if(c[1] > 500) reach_error()`)
+
+Note that only program elements that are (transitively) reachable from main() are checked against any violation of the above criteria, i.e. a program containing unsupported elements that are not utilized is handled correctly. This is necessary for handling most preprocessed (.i) files, as the standard library includes a lot of complex and unsupported code. 
+
+## XCFA formalism
+
+An XCFA is a process- and procedure-based collection of directed graphs (`V`, `L`, `E`) with
+
+* variables `V = {v1, v2, ..., vn}`,
+* locations `L`, with dedicated initial (`l0`), final (`lf`) and error (`le`) locations,
+* edges `E` between locations, labeled with statements over the variables. Statements can be
+    * assignments of the form `v := expr`, where `expr` is an expression with the same type as `v`,
+    * assumptions of the form `assume expr`, where `expr` is a Boolean expression,
+    * havocs of the form `havoc v`,
+    * boundaries of atomic blocks `AtomicBegin`, `AtomicEnd`,
+    * memory operation primitives `Load`, `Store` with optional annotation of `atomic @ordering` where `ordering` is a
+      memory ordering primitive,
+    * call statements of the form `call proc` where `proc` is a referenced procedure (by name).
+
+### Textual representation (DSL)
+
+An example XCFA realizing a two-threaded counter:
+
```

Comment:
I would like to see the used statements explained. There are the standard assume, assign, havoc, but the documentation does not state anything about the meaning of `<-`, `->` or the construct `atomic` and `@relaxed` (what do they mean, what other options are available). This README should provide a complete description of the user facing part of XCFA.